Introduce KSCrashBacktrace

naftaly opened the pull request:
# Summary
This pull request adds the new KSCrashBacktrace module to KSCrash and makes it publicly accessible. The primary goal is to enable other developers to obtain stack traces on demand, independent of a crash event.

## Key Changes
Introduced the ks_backtrace module for capturing stack traces.
Made the module’s functionality public so it can be used outside the crash reporter.
Enables clients to programmatically obtain backtraces whenever needed.

## Motivation
Previously, stack trace functionality was internal to crash handling. By exposing KSCrashBacktrace as a public module, this change empowers developers and other modules to capture stack traces for diagnostics, debugging, or custom monitoring at any time.

GLinnik21 commented on Sources/KSCrashBacktrace/KSBacktrace.m:
```diff
@@ -0,0 +1,64 @@
+//
+//  Created by Alexander Cohen on 2025-05-27.
+//
+//  Copyright (c) 2025 Alexander Cohen. All rights reserved.
```
I think we should preserve Karl’s copyright notice in every file.

naftaly replied:
Sure no problem, wasn't sure to do there.

GLinnik21 commented on Sources/KSCrashBacktrace/include/KSBacktrace.h:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @note This function is not async-signal-safe and must not be called from a signal handler.
+ */
+int ks_backtrace(pthread_t _Nonnull thread, uintptr_t *_Nonnull addresses, int count);
```
Since this is a public module, the Swift API should also feel idiomatic. `ks_backtrace` isn’t a very Swift-friendly name — something like `NS_SWIFT_NAME(captureBacktrace(thread:maxFrames:))` would be more appropriate.

naftaly replied:
Sounds good.

GLinnik21 replied:
Alternatively, we could even consider creating an Objective-C class to represent the backtrace — for example, one that holds the captured addresses. That way, it could be bridged into Swift in a much more idiomatic and Swifty way, instead of dealing with raw C arrays.

naftaly replied:
Anyone using this needs to be "to the metal", so the less layers the better. In swift, this is really easy to use:

```
let entries = 512
var addresses: [UInt] = Array(repeating: 0, count: entries)
let frameCount = captureBacktrace(thread: thread, addresses: &addresses, maxFrames: Int32(entries))
```

GLinnik21 commented on Sources/KSCrashBacktrace/include/KSBacktrace.h:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @note This function is not async-signal-safe and must not be called from a signal handler.
+ */
+int ks_backtrace(pthread_t _Nonnull thread, uintptr_t *_Nonnull addresses, int count);
```
Is `pthread_t` sufficient for this use case, or would it make sense to support the Mach thread API as well?

naftaly replied:
Anyone using this needs to be "to the metal", so the less layers the better. In swift, this is really easy to use:

```
let entries = 512
var addresses: [UInt] = Array(repeating: 0, count: entries)
let frameCount = captureBacktrace(thread: thread, addresses: &addresses, maxFrames: Int32(entries))
```

GLinnik21 reviewed (commented):
So if I understand correctly, the main difference from the system `backtrace()` is that `ks_backtrace` can capture any thread, while `backtrace()` is limited to the current thread.

Are there any places inside KSCrash where we could take advantage of that? For example, I’m looking at `kscm_reportUserException`, where we currently capture only the current thread — maybe this could be a candidate?

https://github.com/kstenerud/KSCrash/blob/c0177658fc771ae833b6d3b8f5f3f5a14d59ec03/Sources/KSCrashRecording/Monitors/KSCrashMonitor_User.c#L52-L63

naftaly commented:
> So if I understand correctly, the main difference from the system `backtrace()` is that `ks_backtrace` can capture any thread, while `backtrace()` is limited to the current thread.
> 
> Are there any places inside KSCrash where we could take advantage of that? For example, I’m looking at `kscm_reportUserException`, where we currently capture only the current thread — maybe this could be a candidate?
> 
> https://github.com/kstenerud/KSCrash/blob/c0177658fc771ae833b6d3b8f5f3f5a14d59ec03/Sources/KSCrashRecording/Monitors/KSCrashMonitor_User.c#L52-L63

What I did was wrap up the private internal cursor API and make it available for use publicly. What ks_backtrace does internally is exactly what all KSCrash code does everywhere. It gets a machine context inits a cursor based on X, and then write that cursor to some output by iterating it. The API I built simply wraps that and stores the cursor iteration in order to store the addresses of the trace. So basically, KSCrash already uses all this internal, I just made it available externally in a simple to use manner.

bamx23 commented on Tests/KSCrashBacktraceTests/KSBacktraceTests.m:
```diff
@@ -0,0 +1,56 @@
+        uintptr_t addresses[frameCount] = { 0 };
+        size_t count = ks_backtrace(thread, addresses, frameCount);
+
+        [expectation fulfill];
```
Let's assert count is not zero. 

Also, could we reliably validate the content of addresses array somehow? At least that `addresses[count - 1] != 0`.

naftaly replied:
I added a few asserts to validate that we're getting some data back.

bamx23 commented on Sources/KSCrashBacktrace/include/KSBacktrace.h:
```diff
@@ -0,0 +1,45 @@
+ */
+
+FOUNDATION_EXPORT
+int ks_backtrace(pthread_t _Nonnull thread, uintptr_t *_Nonnull addresses, int count)
```
Should we call it `ks_captureBacktrace` to make it more consistent with Swift?

bamx23 commented on Sources/KSCrashBacktrace/include/KSBacktrace.h:
```diff
@@ -0,0 +1,45 @@
+
+FOUNDATION_EXPORT
+int ks_backtrace(pthread_t _Nonnull thread, uintptr_t *_Nonnull addresses, int count)
+    NS_SWIFT_NAME(captureBacktrace(thread:addresses:maxFrames:));
```
Don't think we should change the name of this parameter in Swift, as it's present in documentation and described there.
```suggestion
    NS_SWIFT_NAME(captureBacktrace(thread:addresses:count:));
```

bamx23 reviewed (commented):
Why do we need a dedicated new module for this function? Do you imagine more code added here in the future?

I suggest adding it to existing recording module and also providing an ObjC wrapper method that would store the result addresses in an NSArray, which would be a better Swift API either. 

And a related question: is it possible to use NSThread as an argument there (does it have a way to get underlying pthread)?

naftaly commented:
@bamx23 @GLinnik21 what do you think of the latest changes? I think we're in a good spot.

GLinnik21 commented:
Looks good overall — thanks for the updates!

One small thought: since the new module still depends on `KSCrashRecordingCore`, we’re not really gaining a smaller or more lightweight dependency surface. So unless the goal is long-term modularity, maybe exposing something like `KSCrash.captureBacktrace(...)` directly could be the simpler path for now?

bamx23 commented:
@naftaly the main concern from my side is that this one function doesn't look like a big enough use case to make a new module that we'll need to maintain moving forward (at least until v3).

What if instead you just add it into `KSCrashRecordingCore`? The function is a pretty low level use case (C API) and seems a good fit there. And while KSCrash consider `KSCrashRecordingCore` being "internal" API, we still won't break it without major version change, so it's safe to depend on core directly. In my opinion, there are many other functions in `KSCrashRecordingCore` that follow the same idea of being useful helpers for those who "need to go deeper".

naftaly commented:
> @naftaly the main concern from my side is that this one function doesn't look like a big enough use case to make a new module that we'll need to maintain moving forward (at least until v3).
> 
> What if instead you just add it into `KSCrashRecordingCore`? The function is a pretty low level use case (C API) and seems a good fit there. And while KSCrash consider `KSCrashRecordingCore` being "internal" API, we still won't break it without major version change, so it's safe to depend on core directly. In my opinion, there are many other functions in `KSCrashRecordingCore` that follow the same idea of being useful helpers for those who "need to go deeper".

Sounds good, happy to do that. I'll make the changes in a few.

naftaly commented:
I removed the Backtrace module and moved the code to RecordingCore. I also added a symbolication method that allows us to have everything needed for a backend to correctly symbolicate an address from the stack.

@GLinnik21 @bamx23

GLinnik21 commented on Tests/KSCrashRecordingCoreSwiftTests/KSBacktraceTests.swift:
Great! The first Swift unit test is in — time to invite SwiftFormat to the party

naftaly replied:
I ran swift format. I'll leave it up to you folks to decide how you want to setup it up if you do.

GLinnik21 commented on Sources/KSCrashRecordingCore/include/KSBacktrace.h:
```diff
@@ -0,0 +1,98 @@
+ *             It may also briefly suspend the target thread while unwinding its stack.
+ */
+
+int ks_captureBacktrace(pthread_t _Nonnull thread, uintptr_t *_Nonnull addresses, int count)
```
I think `ksbt_` would follow better established pattern - other modules in KSCrashRecordingCore use module-specific prefixes like `ksdl_`, `ksbic_`, `ksmc_`, etc.

naftaly replied:
done

GLinnik21 commented on Sources/KSCrashRecordingCore/include/KSBacktrace.h:
```diff
@@ -0,0 +1,98 @@
+    uint64_t imageSize;
+} CF_SWIFT_NAME(SymbolInformation);
+
+/**
+ * Resolves symbol information for a given instruction address.
+ *
+ * @param address  The instruction address to symbolize.
+ * @param result   A pointer to a KSSymbolInformation structure to be populated. Must not be NULL.
+ *
+ * @return @c true if symbolication succeeded and @c result is populated, @c false otherwise.
+ *
+ * @discussion On success, @c result will contain the symbol name, symbol address, image name,
+ *             image load address, image size, and image UUID associated with @c address.
+ */
```
Should we mention that `ksbic_init` should be called beforehand, or should we call it inside?

naftaly replied:
I'll call it inside.

naftaly replied:
called inside.

GLinnik21 commented on .swiftpm/xcode/xcshareddata/xcschemes/KSCrash-Package.xcscheme:
```diff
@@ -105,6 +105,20 @@
                ReferencedContainer = "container:">
             </BuildableReference>
          </BuildActionEntry>
+         <BuildActionEntry
+            buildForTesting = "YES"
+            buildForRunning = "YES"
+            buildForProfiling = "YES"
+            buildForArchiving = "YES"
+            buildForAnalyzing = "YES">
+            <BuildableReference
+               BuildableIdentifier = "primary"
+               BlueprintIdentifier = "BootTimeMonitor"
+               BuildableName = "BootTimeMonitor"
+               BlueprintName = "BootTimeMonitor"
+               ReferencedContainer = "container:">
+            </BuildableReference>
+         </BuildActionEntry>
```
Is that needed?

naftaly replied:
I don't know, I think it's automatic scheming, but I'm not sure. It did however change. Should I instead add it to the ignore list?

naftaly replied:
reverted those changes.

GLinnik21 commented on .swiftpm/xcode/xcshareddata/xcschemes/KSCrash-Package.xcscheme:
```diff
@@ -228,6 +242,26 @@
          </TestableReference>
+         <TestableReference
+            skipped = "NO">
+            <BuildableReference
+               BuildableIdentifier = "primary"
+               BlueprintIdentifier = "KSCrashBacktraceTests"
+               BuildableName = "KSCrashBacktraceTests"
+               BlueprintName = "KSCrashBacktraceTests"
+               ReferencedContainer = "container:">
+            </BuildableReference>
```
Seems this doesn’t exist anymore

naftaly replied:
same here.

GLinnik21 commented on Sources/KSCrashRecordingCore/KSBinaryImageCache.c:
```diff
@@ -125,25 +126,29 @@ static void ksbic_removeImageCallback(const struct mach_header *header, intptr_t
 }
 
-static bool g_initialized = false;
+static _Atomic(bool) g_Initialized = false;
```
Minor: `g_initialized` follows snake case better

GLinnik21 reviewed (approved):
Fix compilation errors and we’re good to go!

naftaly commented:
> Fix compilation errors and we’re good to go!

What compilation errors are you seeing?

naftaly commented on Sources/KSCrashRecordingCore/KSBacktrace.c:
```diff
@@ -0,0 +1,98 @@
+    result->symbolAddress = info.dli_saddr;
+    result->symbolName = info.dli_sname;
+    result->imageName = info.dli_fname;
+    result->imageAddress = info.dli_fbase;
```
cleaned up the types and did some casting when required.

naftaly commented:
@GLinnik21 @bamx23 Thanks so much for helping get this PR through. I really appreciate the diligent reviews. 🚀
